src/System.Private.CoreLib/src/System/Utf8String.cs

```diff
@@ -0,0 +1,80 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    // This is an experimental type and not referenced from CoreFx but needs to exists and be public so we can prototype in CoreFxLab.
+    public sealed class Utf8String
+    {
+        // Do not reorder these fields. Must match layout of Utf8StringObject in object.h.
+        private readonly int _length;
+        private byte _firstByte;  // would like to mark "readonly" but need to take ref to it in our magic "constructor" that C# doesn't recognize as a constructor.
+
+        private Utf8String() { } // Suppress creation of the public constructor. No one actually calls this.
+
+        public int Length => _length;
+        public ref readonly byte GetPinnableReference() => ref _firstByte;
+
+        // Utf8String constructors
+        // These are special. The implementation methods for these have a different signature from the
+        // declared constructors.
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public extern Utf8String(ReadOnlySpan<byte> value);
+
+#if PROJECTN
+        [DependencyReductionRoot]
+#endif
+#if !CORECLR
+        static
+#endif
+        private Utf8String Ctor(ReadOnlySpan<byte> value)
+        {
```

Comment:
Do we plan to specialize empty strings like they are specialized in regular strings?